File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, right = 0, maxLen = 0;
        vector<int> map(256, -1);
        while(right < s.size()){
            int currChar = s[right];
            if(map[currChar] != -1){
                left = max(left, map[currChar] + 1);
            }

            map[currChar] = right;
            maxLen = max(maxLen, right - left + 1);
            right++;
        }
        return maxLen;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        freq_map = [-1] * 256
        left, right = 0, 0
        max_window = 0
        while right < len(s):
            index = ord(s[right])
            if freq_map[index] != -1:
                left = max(freq_map[index] + 1, left)
            
            freq_map[index] = right
            if max_window < right - left + 1:
                max_window = right - left + 1
            
            right += 1
        
        return max_window

